### ISSUE TITLE
PHP Fatal Error: Nesting level too deep - recursive dependency?

### ISSUE THREAD

**Issue by callmez** [yiisoft/yii2#8348]
```
\Yii::$app->getDb()->createCommand('CREATE TABLE IF NOT EXISTS `test` (type INT(11))')->execute();
$models = [
    new \app\models\Test(['type' => 0]),
    new \app\models\Test(['type' => 0]),
];
\yii\helpers\ArrayHelper::multisort($models, 'type', SORT_ASC);
```
Test  model file
```
<?php
namespace app\models;

use yii\db\ActiveRecord;
use yii\behaviors\TimestampBehavior;

class Test extends ActiveRecord
{
    public static function tableName()
    {
        return 'test';
    }

    public function behaviors()
    {
        return [
            'timestamp' => TimestampBehavior::className()
        ];
    }
}
```

Get error

```
PHP Fatal Error – yii\base\ErrorException

Nesting level too deep - recursive dependency?
```

**Comment by callmez** [comment]
related https://github.com/yiisoft/yii/issues/2082

**Comment by cebe** [comment]
Reproduced. Only happens if timestamp behavior is added. The error happens inside of PHP so it is hard to figure out the cause...

**Comment by cebe** [comment]
Reproduced with plain PHP:

```php
<?php

class Related
{
	public $owner;
}

class Model
{
	public $related;
	public $type = 0;

	public function __construct($type)
	{
		$this->related = new Related();
		$this->related->owner = $this;

		$this->type = $type;
	}
}

$models = array(
	0 => new Model(1),
	1 => new Model(1),
);

$args = [
	[0 => 1, 1 => 1],
	SORT_ASC,
	SORT_REGULAR,
];
$args[] = &$models;

var_dump(call_user_func_array('array_multisort', $args));
print_r($models);
```

As far as I see the error results from the fact that the given ordering does not result in an ordering so PHP will compare the objects directly and this results in the error:

```php
var_dump($models[0] < $models[1]);
```

Seems that the fix is to force a default ordering if given columns do not specify an ordering:
https://github.com/yiisoft/yii/issues/2082#issuecomment-38397101
need to verify this.

**Comment by Bhoft** [comment]
the same also happens also with the AttributeBehavior.
are they any news how to fix this issue or when it will be resolved?

**Comment by zebratuk** [comment]
Problem can be solved by adding `__toString` into `yii\base\Object` and adding `$args[] = SORT_STRING;` after https://github.com/yiisoft/yii2/blob/master/framework/helpers/BaseArrayHelper.php#L448
But this solution good only for objects, not for all cases.

**Comment by beowulfenator** [comment]
The same problem is with `Zelenin\yii\behaviors\Slug`. Also, i've noticed that this error occurs only with a certain amount of entries in the array - on small arrays everything goes through fine.

**Comment by beowulfenator** [comment]
Updating PHP to 5.5.30 didn't help.

**Comment by niektenhoopen** [comment]
@beowulfenator Just to be sure: I can somewhat recall having this error message on a dataset with a left join, that contained null values (non-existing records for the relation), and then performing array_multisort() on this array. Like I said, just to be sure, could this be the case with your issue as well?  (Your "on small arrays everything goes fine" reminded me about this issue I had, because my small datasets did not contain these null values)

**Comment by beowulfenator** [comment]
Wow! Thanks, @niektenhoopen, this is exactly my case. The models do have some null attributes and removing them seems to fix the problem.

**Comment by m8rge** [comment]
I resolve this using http://www.php.net/manual/en/function.array-multisort.php#113445 .

**Comment by andrewnester** [comment]
Hi to all! 
I've just created PR with fix for it, anyone interested are welcome to check it

**Comment by SilverFire** [comment]
Fixed in 78de315
